# D209 Task 2 PA (Predictive Analysis)
Vincent Taylor

---

## Part I

***Research Question***

(A1): The research question that I will be analyzing is what factors  contribute most to `tenure`. This question is the same from my D208 assessment as I believe that delving deeper into this specific question would create value for our organization. The benefits associated with this analysis are being able to understand factors that contribute to customer tenure. By finding the factors that contribute we can then potentially identify specific products or create incentives in order to maintain more long-term customers. As previously discussed in my prior markup finding attributes that predict tenure can lead to a direct cost savings by not having to focus as heavily on marketing to new customers. I will answer this question using decision trees. 

**Objectives and Goals for this analysis**

Using DecisionTreeRegressor from the sklearn library, the primary objective of this analysis is to look into the variables that could affect a customer's tenure. Our main goal will be to identify the dataset's most significant characteristics that may have an influence on tenure. For this reason, we are going to utilize the same set of features that were included in the original d208 MLR model for task1.These variables will then be preprocessed to fit our decision tree regression model. Our ultimate goal is to identify potential customer churning points and facilitate proactive customer service actions. Using predictive modeling helps strengthen our customer service and helps us target potential churners and tailor products and services to keep customers long term. In addition, by recognizing these important factors, we can more effectively allocate funds and resources, strengthening our position as industry leaders.


---

## Part II Method Justification

***Justificaton of Method***

For this task I will be using decisions trees. I chose this method due to its ease and convience of being able to visulaize results and the fact it is more easy to interpret. Decision trees may also be used with both numerical and categorical data. What makes decision trees different from the random forest is that decision trees themselves are graphs that show all of the potential outcomes of a decision using the branches to represent different paths. Another reason I chose this compared to the random forest method is that the final model for a decision tree can be visualized. A downside to decision trees is that it is prone to overfitting,and innacurate final results. Due to this a random forest may be a more suitable choice due to it being able to avoid overfitting by using multiple trees, another downside is that in most real world cases the better method to use would be the random forest algorithm due to modern computers being able to handle larger datasets.

***How Decision Trees Work***

Decision trees use multiple algorithms to decide where to split a node into, esentially a step by step process until we can arrive at answer we are looking for. For example we are trying to predict customer tenure, the tree starts by taking all of the features we have such "does a customer have online_security? How much bandwidth do they use? etc, and based on the all the different types of responses it creates a new branch until you reach the specific leaf that tells you the answer. Furthermore Analytics Vidhya goes on to describe the way decision trees work through recursive partitioning where each node is divided into child nodes and the process continues until a stopping criterion is met.


***Assumption of Classification Method***

One assumption of decision tree models is that, there is an assumption of independence among features. This means that the model treats each feature seperately when making decisions. Analytics Vidhya goes on to state that this may not actually hold to be true in actual cases as a decision tree can perform well even if the features show high correlation.

***Package/Libarary Selection**

I chose to use Python to perform this analysis due to its ability to handle data analysis as well as it's ease of use when it comes to visualizing and performing statitsitcal modeling. I will also be using various packages in order to perform this analysis.

- Numpy and Pandas: Data analysis and preperation
- Matplotlib and Seaborn:Data visualization
- Scikit-learn: Machine learning and predictive modeling

---

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import r2_score, accuracy_score, mean_squared_error as mse
from sklearn.tree import DecisionTreeRegressor
from sklearn import tree 

# import data from csv
df = pd.read_csv('churn_clean.csv', index_col=0)
df.head()

Customer_id                           Interaction  \
CaseOrder                                                     
1             K409198  aa90260b-4141-4a24-8e36-b04ce1f4f77b   
2             S120509  fb76459f-c047-4a9d-8af9-e0f7d4ac2524   
3             K191035  344d114c-3736-4be5-98f7-c72c281e2d35   
4              D90850  abfa2b40-2d43-4994-b15a-989b8c79e311   
5             K662701  68a861fd-0d20-4e51-a587-8a90407ee574   

                                        UID         City State  \
CaseOrder                                                        
1          e885b299883d4f9fb18e39c75155d990  Point Baker    AK   
2          f2de8bef964785f41a2959829830fb8a  West Branch    MI   
3          f1784cfa9f6d92ae816197eb175d3c71      Yamhill    OR   
4          dc8a365077241bb5cd5ccd305136b05e      Del Mar    CA   
5          aabb64a116e83fdc4befc1fbab1663f9    Needville    TX   

                          County    Zip       Lat        Lng  Population  ...  \
CaseOrder                                                                 ...   
1          Prince of Wales-Hyder  99927  56.25100 -133.37571          38  ...   
2                         Ogemaw  48661  44.32893  -84.24080       10446  ...   
3                        Yamhill  97148  45.35589 -123.24657        3735  ...   
4                      San Diego  92014  32.96687 -117.24798       13863  ...   
5                      Fort Bend  77461  29.38012  -95.80673       11352  ...   

          MonthlyCharge Bandwidth_GB_Year Item1  Item2  Item3  Item4 Item5  \
CaseOrder                                                                    
1            172.455519        904.536110     5      5      5      3     4   
2            242.632554        800.982766     3      4      3      3     4   
3            159.947583       2054.706961     4      4      2      4     4   
4            119.956840       2164.579412     4      4      4      2     5   
5            149.948316        271.493436     4      4      4      3     4   

          Item6 Item7  Item8  
CaseOrder                     
1             4     3      4  
2             3     4      4  
3             3     3      3  
4             4     3      3  
5             4     4      5  

[5 rows x 49 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 1 to 10000
Data columns (total 49 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Customer_id           10000 non-null  object 
 1   Interaction           10000 non-null  object 
 2   UID                   10000 non-null  object 
 3   City                  10000 non-null  object 
 4   State                 10000 non-null  object 
 5   County                10000 non-null  object 
 6   Zip                   10000 non-null  int64  
 7   Lat                   10000 non-null  float64
 8   Lng                   10000 non-null  float64
 9   Population            10000 non-null  int64  
 10  Area                  10000 non-null  object 
 11  TimeZone              10000 non-null  object 
 12  Job                   10000 non-null  object 
 13  Children              10000 non-null  int64  
 14  Age                   10000 non-null  int64  
 15  Income             

### Data Cleaning Goals

(C1)
In order to prepare the data I will need to encode the categorical variables to numeric values,  I will also need to change the columns to a pythonic naming convention as done in prior tasks, I will also need to check to ensure that there are no null values in the dataset, lastly will will need to make the dataset more concise and remove features that we arent using in the analysis in order to keep memory stress low on our machine while also ensure we are extracting meaningful features for this analysis. I will also be using features identified in d208 task 1 for this model, with the goal of 70% accuracy at  minimum. 



### Variable Selection

(C2)
Below are features I intend to use for this decision tree model. The feature selection process I used is taking the features that showed promise in d208 task2. I will be removing unecessary variables in order to reduce dimensonality as previously discussed. X represents the independent variables and y represents our dependent variable. 

---
|Variable|Continous/Categorical|x/y|
|---|---|---|
|Tenure|Continuous|***Dependent***|
|bandwidth_gb_year|Continuous|Independent|
|monthly_charge|Continuous|Independent|
|age|Continuous|Independent|
|children|Continuous|Independent|
|online_security|Categorical|Independent|

---

### Exploratory Data Analysis

(C3)
Once the inital data cleaning has been completed I will be exploring the dataset using functions from prior tasks such as .info(), .describe, as well as .value_counts(). I will also verify the dataset to ensure that encoding has been completed as well as using .corr() function in order to check for any relationships between our y variable and the independent x variables selected for classification. Since we are using DecisionTreeRegressor feature scaling will not be required.

### Feature Scaling

(C3)
As previously mentioned feature scaling  of numeric values is not required for DecisionTree models. Scaling features doesn't affect the structure or decision rules of the tree. The relative relationships and ordering between different feature values remain the same after scaling. As decision trees split data based on ranks and orders, not absolute values, scaling doesn't alter the decision boundaries.


In [3]:
#Data Cleaning

#renaming columns to python casing.
col_head = {
    'CaseOrder': 'case_order',
    'Customer_id': 'customer_id',
    'Interaction': 'interaction',
    'City': 'city',
    'State': 'state',
    'County': 'county',
    'Zip': 'zip',
    'Lat': 'lat',
    'Lng': 'lng',
    'Population': 'population',  
    'Area': 'area',  
    'Job': 'job',
    'Children': 'children',
    'Age': 'age', 
    'Income': 'income',  
    'Marital': 'marital_type',
    'Gender': 'gender',
    'Churn': 'churn',
    'Outage_sec_perweek': 'outage_sec_perweek',
    'Email': 'email',
    'Contacts': 'contacts',
    'Yearly_equip_failure': 'yearly_equip_failure',
    'Techie': 'techie',
    'Contract': 'contract',
    'Port_modem': 'port_modem',
    'Tablet': 'tablet',
    'InternetService': 'internet_service',
    'Phone': 'phone',
    'Multiple': 'multiple_lines',
    'OnlineSecurity': 'online_security',
    'OnlineBackup': 'online_backup',
    'DeviceProtection': 'device_protection',
    'TechSupport': 'tech_support',
    'StreamingTV': 'streaming_tv',
    'StreamingMovies': 'streaming_movies',
    'PaperlessBilling': 'paperless_billing',
    'PaymentMethod': 'payment_method',
    'Tenure': 'tenure',
    'MonthlyCharge': 'monthly_charge',
    'Bandwidth_GB_Year': 'bandwidth_gb_year',
    'Item1': 'timely_response',
    'Item2': 'timely_fixes',
    'Item3': 'timely_replacements',
    'Item4': 'reliability',
    'Item5': 'options',
    'Item6': 'respectful_response',
    'Item7': 'courteous_exchange',
    'Item8': 'active_listening'}
df.rename(columns=col_head, inplace=True)

#Reexpressing the values of payment method to be more concise.
dict_pay = {"payment_method": {
     "Credit Card (automatic)": "Autopay",
     "Bank Transfer(automatic)": "Autopay",
     "Electronic Check": "eCheck",
     "Mailed Check": "Check",
}}
df.replace(dict_pay, inplace = True)

# dropping columns not relevant to analysis
df = df.drop(columns=['customer_id', 'interaction', 'UID', 'city', 'state', 'county', 'zip',
       'lat', 'lng', 'population', 'area', 'TimeZone', 'job',
       'income', 'marital_type', 'gender',
       'outage_sec_perweek', 'email', 'contacts', 'yearly_equip_failure',
       'techie', 'contract', 'port_modem', 'tablet', 'internet_service',
       'phone', 'multiple_lines', 'online_backup',
       'device_protection', 'tech_support', 'streaming_tv', 'streaming_movies',
       'paperless_billing', 'payment_method', 'timely_response', 'timely_fixes',
       'timely_replacements', 'reliability', 'options', 'respectful_response',
       'courteous_exchange', 'active_listening', 'churn'])

In [4]:
#Checking for null values
if df.isnull().sum().sum() == 0:
    print("No null values.")
else:
    print("Null values exist!")

No null values.


In [5]:
#Creating dictionary for unique values and total count for each variable in dataset.
unique_count = {}

for column in df.columns:
    unique_value = df[column].unique()
    counts = df[column].value_counts()
    unique_count[column] = (unique_value, counts)

In [6]:
unique_count['online_security']

(array(['Yes', 'No'], dtype=object),
 No     6424
 Yes    3576
 Name: online_security, dtype: int64)

### Summary Statistics

*Note variables will not need to be scaled.*

---

In [7]:
df.describe()

children           age        tenure  monthly_charge  \
count  10000.0000  10000.000000  10000.000000    10000.000000   
mean       2.0877     53.078400     34.526188      172.624816   
std        2.1472     20.698882     26.443063       42.943094   
min        0.0000     18.000000      1.000259       79.978860   
25%        0.0000     35.000000      7.917694      139.979239   
50%        1.0000     53.000000     35.430507      167.484700   
75%        3.0000     71.000000     61.479795      200.734725   
max       10.0000     89.000000     71.999280      290.160419   

       bandwidth_gb_year  
count       10000.000000  
mean         3392.341550  
std          2185.294852  
min           155.506715  
25%          1236.470827  
50%          3279.536903  
75%          5586.141370  
max          7158.981530

 `age` has 10,000 counts and has a min of 18 with a max of 89. The average age of a customer is 53.

 `tenure` has 10,000 counts with a minimum of 1 month and a maximum of 72 months or 6 years. On average our customers stay with the company for 2.94 years or 34 months.

 `monthly_charge` has 10,000 counts with a minimum cost of 80 dollars and a maximum cost of 290 dollars. on average our customers pay 172 dollars a month for our service.

`bandwidth_gb_year` has 10,000 counts with a minimum usage of 155 and a maximum of 7158 gb used. On average our customers use 3392 gb per year.


In [8]:
# Converting yes/no column to 1/0
yes_no_col = ['online_security']
df[yes_no_col] = df[yes_no_col].replace({'Yes': 1, 'No': 0})

In [9]:
df.corr().abs()

children       age  online_security    tenure  \
children           1.000000  0.029732         0.006937  0.005091   
age                0.029732  1.000000         0.011213  0.016979   
online_security    0.006937  0.011213         1.000000  0.002508   
tenure             0.005091  0.016979         0.002508  1.000000   
monthly_charge     0.009781  0.010729         0.047734  0.003337   
bandwidth_gb_year  0.025585  0.014724         0.021006  0.991495   

                   monthly_charge  bandwidth_gb_year  
children                 0.009781           0.025585  
age                      0.010729           0.014724  
online_security          0.047734           0.021006  
tenure                   0.003337           0.991495  
monthly_charge           1.000000           0.060406  
bandwidth_gb_year        0.060406           1.000000

In [10]:
df.head()

children  age  online_security     tenure  monthly_charge  \
CaseOrder                                                              
1                 0   68                1   6.795513      172.455519   
2                 1   27                1   1.156681      242.632554   
3                 4   50                0  15.754144      159.947583   
4                 1   48                1  17.087227      119.956840   
5                 0   83                0   1.670972      149.948316   

           bandwidth_gb_year  
CaseOrder                     
1                 904.536110  
2                 800.982766  
3                2054.706961  
4                2164.579412  
5                 271.493436

### Provide Copy of cleaned dataset

---

In [11]:
# Save the entire dataset (C4)
df.to_csv('d209_cleaned_task2.csv', index=False)

### Part IV

---

In [12]:
# Splitting data into train and test data (D1)
X = df[['children', 'age', 'online_security',
       'monthly_charge', 'bandwidth_gb_year']]

y = df.tenure

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

***(D2)***

> I am using the 'test_train_split' function in the scikit learn library in order to split the testing and training data. I split the dataset into both training and testing data using a 70/30 split. The 30% will be the training data that will be used to help the model learn. 

In [13]:
#Saving four datasets to CSV (D1)
X_train.to_csv('t2_d209_X_train.csv')
X_test.to_csv('t2_d209_X_test.csv')
y_train.to_csv('t2_d209_y_train.csv')
y_test.to_csv('t2_d209_y_test.csv')

In [14]:
# Define parameters
params_dt = {
    'max_depth': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    'min_samples_leaf': [10, 25, 50, 75, 100]
}

# Create regressor 
dt = DecisionTreeRegressor(random_state=42)

# Perform GridSearchCV
grid_dt = GridSearchCV(estimator=dt, param_grid=params_dt, scoring='r2', cv=10, n_jobs=-1)

# Fit GridSearch to training data
grid_dt.fit(X_train, y_train)

# Get the best estimator from the grid search
best_model = grid_dt.best_estimator_

# Predict on the test set
y_pred = best_model.predict(X_test)

# Calculate and print test set metrics
test_mse = mse(y_test, y_pred)
print('Test set Mean Squared Error: {:.3f}'.format(test_mse))

test_r2 = r2_score(y_test, y_pred)
print('Test set R-squared: {:.3f}'.format(test_r2))

# Display best parameters
print(grid_dt.best_params_)

Test set Mean Squared Error: 7.700
Test set R-squared: 0.989
{'max_depth': 9, 'min_samples_leaf': 25}


***(D2) cont***


> The code I've built takes a structured approach to optimize the DecisionTreeRegressor model. Initially, I chose a range of hyperparameters, such as 'max_depth' and 'min_samples_leaf', to explore variations in tree depth and leaf node sample requirements. I then used GridSearchCV in order to identify the optimal set that maximizes the model's predictive performance.

> After hyperparameter tuning, I used best_estimator to find the best model fit. best_estimator takes the highest R-squared score during cross-validation. After that, I utilize the 'best_model' to predict customer tenure on the test set, allowing me to assess its predictive accuracy. Metrics like Mean Squared Error (MSE) and R-squared are computed to quantify the model's precision. Using this approach optimizes the decision tree's structure but also ensures the selection of the best parameters, which builds the model's predictive capability on unseen data.

In [15]:
#Building final tuned model

#instantiate final tuned model with best parameters
final_dt = DecisionTreeRegressor(max_depth=9, min_samples_leaf=25, random_state=1)

#fitting data to training set
final_dt.fit(X_train, y_train)

# predicting test set labels
y_pred = final_dt.predict(X_test)

# printing mean squared error
test_mse = mse(y_test, y_pred)
print('Test set MSE: {:.3f}'.format(test_mse))

#printing r-squared score
test_r2 = r2_score(y_test, y_pred)
print('Test set R-squared: {:.3f}'.format(test_r2))

Test set MSE: 7.700
Test set R-squared: 0.989


### Part V

---

#### (E1) MSE and R-squared 

The test set's MSE (Mean Squared Error) of 7.700 shows that the model's prediction error rate is extremely low. Lower values indicate better accuracy. Moreover the MSE of 7.700 shows that the squared differences between actual and predicted values are very low. I would also need more domain knowledge of the industry as a whole to decide wether this score is high or low, but in this context I am viewing it as low.

---
#### (E2) Prediction Implications

The R-squared of 0.989 shows that the model predicts almost perfectly the variability of the target variable based on the features provided and based on the optimal leaf samples (25) and maximum depth (9). Furthermore a score of 98% indicates that our model explains a majority the variability of the response data around its mean. However based on the decision trees having an overfitting risk I am also hesitant to accept the result right away and would probably recommend more testing before implementing into a production environment, until I am sure I could replicate the results on unseen data.

---
#### (E3) Limitations of analysis

One limitation of the data analysis is the as previously discussed is overfitting. Overfitting occurs when a model learns specific patterns in the training data to the point that it performs very well on the training set but struggles to generalize to new, unseen data. This could be a concern if the model's high performance is solely due to overfitting or noise present in the training data, rather than capturing genuine relationships. 

---
#### (E4) Recommended Course of Action

My recommendation on this analysis would be to validate the model's performance on an unseen dataset, potentially an industry competitor or if we are able to pull historical data, validating the performance against historicals. If the model maintains high accuracy and generalizability across different datasets, it shoud further be implemented into a production environment.

---

### Panopto Recording

https://wgu.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=58b127fd-1759-4dd0-9728-b0d501842c7c

### Sources

Implement Decision Tree in Python using sklearn|Implementing decision tree in python. (n.d.). Www.youtube.com. Retrieved December 9, 2023, from https://www.youtube.com/watch?v=zvFot5vs6aQ

Saini, A. (2021, August 29). Decision Tree Algorithm - A Complete Guide. Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/08/decision-tree-algorithm/#:~:text=Decision%20trees%20often%20assume%20that

The Best Guide On How To Implement Decision Tree In Python. (n.d.). Simplilearn.com. https://www.simplilearn.com/tutorials/machine-learning-tutorial/decision-tree-in-python

Saini, A. (2021, August 29). Decision Tree Algorithm - A Complete Guide. Analytics Vidhya. https://www.analyticsvidhya.com/blog/2021/08/decision-tree-algorithm/#:~:text=Decision%20trees%20use%20a%20recursive


---